### PR TITLE
Feature/game states

### PR DESCRIPTION
**Pull request by David-Hanna**
GAME STATES YAY. Have fun reviewer.

### CODE REVIEW

File: CS483/CS483/Kartaclysm/KartGame.cpp [hunk 1]
```diff
@@ -30,35 +30,19 @@ bool Kartaclysm::KartGame::Init()
...
-	
-
-	mContextParams.insert(std::pair<std::string, std::string>("Player1_KartDefinitionFile", strKartFile));
-	mContextParams.insert(std::pair<std::string, std::string>("Player1_DriverDefinitionFile", strDriverFile));
-	mContextParams.insert(std::pair<std::string, std::string>("Player1_CameraDefinitionFile", strCameraBottomFile));
-
-	mContextParams.insert(std::pair<std::string, std::string>("Light", "CS483/CS483/Kartaclysm/Data/Lights/light.xml"));
-
 	// Setup State Machine and push first state
 	//TODO we need to get rid of the use of hardcoded int ids for states. It's ugly and error prone
```
Comment (bradcooper314) [inline CS483/CS483/Kartaclysm/KartGame.cpp]: You can get rid of this comment now.

File: CS483/CS483/Kartaclysm/StateMachine/Gameplay/GameplayStatesEnum.h [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+//
+// An enumeration to replace the int values in the state machine.
+//========================================================================
+
+#ifndef GAMEPLAY_STATES_ENUM
+#define GAMEPLAY_STATES_ENUM
+
+namespace Kartaclysm
+{
+	enum GameplayStates
```
Comment (bradcooper314) [inline CS483/CS483/Kartaclysm/StateMachine/Gameplay/GameplayStatesEnum.h]: Consider putting this enum inside of another namespace as well, similar to InputActions.h, and maybe moving it into GamePlayState.h as they are tightly coupled.

File: CS483/CS483/Kartaclysm/StateMachine/Gameplay/StateMainMenu.cpp [hunk 3]
```diff
@@ -1,105 +1,62 @@
 //------------------------------------------------------------------------
 // StateMainMenu
-// Author:	Bradley Cooper
+// Author:	David Hanna
```
Comment (bradcooper314) [inline CS483/CS483/Kartaclysm/StateMachine/Gameplay/StateMainMenu.cpp:3]: Wow.

File: CS483/CS483/Kartaclysm/StateMachine/Gameplay/StatePlayerSelectionMenu.cpp [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+		m_pGameObjectManager->Update(p_fDelta);
+
+		std::string strKartFile = "CS483/CS483/Kartaclysm/Data/Racer/kart_juggernaut.xml";
+		std::string strDriverFile = "CS483/CS483/Kartaclysm/Data/Racer/driver_clockmaker.xml";
+		std::string strCameraTopFile = "CS483/CS483/Kartaclysm/Data/Camera/camera_top.xml";
+		std::string strCameraBottomFile = "CS483/CS483/Kartaclysm/Data/Camera/camera_bottom.xml";
+
+		std::map<std::string, std::string> mContextParameters;
+		mContextParameters.insert(std::pair<std::string, std::string>("PlayerCount", "2"));
+		mContextParameters.insert(std::pair<std::string, std::string>("Player0_KartDefinitionFile", strKartFile));
```
Comment (bradcooper314) [inline CS483/CS483/Kartaclysm/StateMachine/Gameplay/StatePlayerSelectionMenu.cpp]: The context params are being created and inserted every Update() call even if they're not used. This should be inside the if(Enter key is pressed) statement below.

File: CS483/CS483/Kartaclysm/StateMachine/Gameplay/StateRacing.cpp [hunk 5]
```diff
@@ -124,6 +106,9 @@ void Kartaclysm::StateRacing::Enter(const std::map<std::string, std::string>& p_
 		pTrackComponent->RegisterRacer(vRacers.at(i));
 		vRacers.at(i)->GetTransform().TranslateXYZ(1.0f * i, 0.0f, 0.0f); // TODO: Better positioning
 	}
+
+	// TODO: WHY IS PRINTING THE GAME OBJECT MANAGER FIXING A BUG????
+	m_pGameObjectManager->Print();
```
Comment (bradcooper314) [inline CS483/CS483/Kartaclysm/StateMachine/Gameplay/StateRacing.cpp:111]: Seems legit.

File: CS483/CS483/Kartaclysm/StateMachine/Gameplay/StateTrackSelectionMenu.cpp [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+			m_pStateMachine->Push(STATE_RACING, m_mContextParameters);
+		}
+		else if (HeatStroke::KeyboardInputBuffer::Instance()->IsKeyDownOnce(GLFW_KEY_UP))
+		{
+			switch (m_iTrackSelection)
+			{
+			case 1:
+				m_iTrackSelection = 0;
+				m_pGameObjectManager->DestroyGameObject(m_pCurrentHighlight);
+				m_pCurrentHighlight = m_pGameObjectManager->CreateGameObject("CS483/CS483/Kartaclysm/Data/Menus/TrackSelectionMenu/track_selection_highlight_noob_zone.xml");
```
Comment (bradcooper314) [inline CS483/CS483/Kartaclysm/StateMachine/Gameplay/StateTrackSelectionMenu.cpp:80]: Surely there is a better way, but its  not an issue for now.

**Review by bradcooper314 — APPROVED** [review]
A lot of printf statements being pushed onto develop. Otherwise, could use more Comic Sans.

File: CS483/CS483/Kartaclysm/StateMachine/Gameplay/StateMainMenu.cpp (hunk 3 above)
Comment (David-Hanna) [reply CS483/CS483/Kartaclysm/StateMachine/Gameplay/StateMainMenu.cpp:3]: ;)

File: CS483/CS483/Kartaclysm/KartGame.cpp (hunk 1 above)
Comment (David-Hanna) [reply CS483/CS483/Kartaclysm/KartGame.cpp]: Done

File: CS483/CS483/Kartaclysm/StateMachine/Gameplay/StatePlayerSelectionMenu.cpp (hunk 4 above)
Comment (David-Hanna) [reply CS483/CS483/Kartaclysm/StateMachine/Gameplay/StatePlayerSelectionMenu.cpp]: Yes. Missed this for some reason. Done.

File: CS483/CS483/Kartaclysm/StateMachine/Gameplay/StateTrackSelectionMenu.cpp (hunk 6 above)
Comment (David-Hanna) [reply CS483/CS483/Kartaclysm/StateMachine/Gameplay/StateTrackSelectionMenu.cpp:80]: States are pretty hard-coded things anyway.

**Pull request merged**